### PR TITLE
Added Socialite provider for Authentik

### PR DESCRIPTION
**Pull request by rf152**
Added a socialite provider for [Authentik](https://goauthentik.io/)

Base_URL is extracted to allow customisation to a specific Authentik instance

### CODE REVIEW

**Comment by AlexZorzi** [comment]
Great! bump PR
We really need this 👍

File: src/Authentik/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+];
+```
+
+### Usage
+
+You should now be able to use the provider like you would regularly use Socialite (assuming you have the facade installed):
+
+```php
+return Socialite::driver('authentik')->redirect();
+```
```
Comment (atymic) [inline src/Authentik/README.md:43]: Can we add returned fields, or at least document how to access the non standard fields?

File: src/Authentik/composer.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "socialiteproviders/authentik",
+    "description": "Authentik OAuth2 Provider for Laravel Socialite",
+    "license": "MIT",
+    "authors": [{
+        "name": "rf152",
+        "email": "git@rf152.co.uk"
+    }],
+    "require": {
+        "php": "^7.2",
```
Comment (atymic) [inline src/Authentik/composer.json:10]:
```suggestion
        "php": "^8.0",
```

File: src/Authentik/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Authentik
+
+```bash
+composer require socialiteproviders/authentik
+```
+
+## Installation & Basic Usage
+
+Please see the [Base Installation Guide](https://socialiteproviders.com/usage/), then follow the provider specific instructions below.
```
Comment (atymic) [inline src/Authentik/README.md:9]: Can we add a link to the authentik docs?

Comment (AlexZorzi) [reply src/Authentik/README.md:9]: leaving here Oauth2 docs for authentik, just in case https://goauthentik.io/docs/providers/oauth2/ :)

**Comment by rf152** [comment]
I've made changes to the README and composer.json, and a couple of changes to the code itself to return more than just the email address.

**Comment by AlexZorzi** [comment]
> I've made changes to the README and composer.json, and a couple of changes to the code itself to return more than just the email address.

Do you retrive all the available information? On authentik it says openid profile and email are the requestable values

**Comment by rf152** [comment]
Yep, and in the profile against the default Laravel User object the attributes that match are email and name, so it maps both of those.

**Comment by AlexZorzi** [comment]
> Yep, and in the profile against the default Laravel User object the attributes that match are email and name, so it maps both of those.

i would also include all the other available information that could be usefull like 'nickname' 'groups' and others

if you add 'email' like so 'protected $scopes = ['openid goauthentik.io/api profile email'];'

you can also get 'email_verified'

Immagine 2022-07-29 173244

**Comment by AlexZorzi** [comment]
still some values missing, refer to this commit as example, also i added a default value in case authentik dosent provide the full access to profile email and openid

https://github.com/AlexZorzi/Providers/commit/927a5c5802e74cf11fe05d409236278b1341a003

**Comment by rf152** [comment]
I've added all the fields that you've mentioned; not sure how it will work if the user model doesn't have those values - I'm guessing it won't throw an error?

**Comment by AlexZorzi** [comment]
> I've added all the fields that you've mentioned; not sure how it will work if the user model doesn't have those values - I'm guessing it won't throw an error?

nice, yes if the key is not present it will default to "" apart for email_verified that as a boolean will default to null

**Comment by rf152** [comment]
> > I've added all the fields that you've mentioned; not sure how it will work if the user model doesn't have those values - I'm guessing it won't throw an error?
> 
> nice, yes if the key is not present it will default to "" apart for email_verified that as a boolean will default to null

I mean the User model within Laravel. I can see that you've set a default for the ones coming in from OAuth.

More that the default User model in Laravel doesn't have half those fields....

**Comment by AlexZorzi** [comment]
I would say that you can always filter out stuff that you don’t need afterwards. I would give the user all the available data coming from authentik anyway 
Maybe @atymic has a better way in mind to be compatible with laravel default User model without loosing data that could be useful

File: src/Authentik/composer.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "socialiteproviders/authentik",
+    "description": "Authentik OAuth2 Provider for Laravel Socialite",
+    "license": "MIT",
```
Comment (lucasmichot) [inline src/Authentik/composer.json:4]: Can you also add the missing keywords section?

Comment (lucasmichot) [reply src/Authentik/composer.json:4]: Also the support section is missing.

File: src/Authentik/Provider.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+use SocialiteProviders\Manager\OAuth2\AbstractProvider;
+use SocialiteProviders\Manager\OAuth2\User;
+
+class Provider extends AbstractProvider
+{
+    /**
+     * Unique Provider Identifier.
+     */
+    const IDENTIFIER = 'AUTHENTIK';
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:13]:
```suggestion
    public const IDENTIFIER = 'AUTHENTIK';
```

File: src/Authentik/Provider.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+     * {@inheritdoc}
+     */
+    public static function additionalConfigKeys()
+    {
+        return ['base_url'];
+    }
+
+    protected function getBaseUrl()
+    {
+        return rtrim($this->getConfig('base_url'), '/');
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:30]: Can you raise an InvalidArgumentException if this config key is no set?

File: src/Authentik/Provider.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        return $this->getBaseUrl().'/application/o/token/';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        $response = $this->getHttpClient()->get($this->getBaseUrl().'/application/o/userinfo/', [
+            'headers' => [
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:55]: Please use the RequestOptions headers const instead

File: src/Authentik/Provider.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                'Authorization' => 'Bearer '.$token,
+            ],
+        ]);
+
+        return json_decode($response->getBody(), true);
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:60]:
```suggestion
        return json_decode((string) $response->getBody(), true);
```

File: src/Authentik/Provider.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            'nickname'           => $user['nickname'] ?? '',
+            'groups'             => $user['groups'] ?? '',
+            'sub'                => $user['sub'] ?? '',
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:84]: Useless as the parent already sets the grant_type

File: src/Authentik/Provider.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function mapUserToObject(array $user)
+    {
+        return (new User())->setRaw($user)->map([
+            'email'              => $user['email'] ?? '',
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:69]: See `\Laravel\Socialite\Contracts\User` for the required fields
You need at mminima an email, and id or a nickname to identify the user.

check for example https://github.com/laravel/socialite/blob/5.x/src/Two/GoogleProvider.php#L72-L78

File: src/Authentik/Provider.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+{
+    /**
+     * Unique Provider Identifier.
+     */
+    const IDENTIFIER = 'AUTHENTIK';
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopes = ['openid goauthentik.io/api profile email'];
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:18]: This scope format looks like an error

**Review by lucasmichot — CHANGES_REQUESTED** [review]

**Review by lucasmichot — CHANGES_REQUESTED** [review]
This provider is missing in `monorepo-builder.yml`

Please double check https://socialiteproviders.com/contribute/#creating-a-provider

File: src/Authentik/Provider.php (hunk 10 above)
Comment (AlexZorzi) [reply src/Authentik/Provider.php:69]: Am I missing something?
In the part of the code you are referring an email and a nickname are already set

Comment (lucasmichot) [reply src/Authentik/Provider.php:69]: Both email and nickname can be empty string.
And no user id is provided.

If you don't provide a unique user id, you need to provide some non-empty unique strings somewhere, right ?

File: src/Authentik/Provider.php (hunk 11 above)
Comment (rf152) [reply src/Authentik/Provider.php:18]: It's a valid scope format. (tested and working)

File: src/Authentik/Provider.php (hunk 10 above)
Comment (rf152) [reply src/Authentik/Provider.php:69]: I've adjusted the returned values to return the sub as an id field. That will never be empty (it's the subject of the authentication)

File: src/Authentik/Provider.php (hunk 9 above)
Comment (rf152) [reply src/Authentik/Provider.php:84]: Removed

File: src/Authentik/Provider.php (hunk 7 above)
Comment (rf152) [reply src/Authentik/Provider.php:55]: I don't know what you mean by this one - are you able to give an example?

File: src/Authentik/Provider.php (hunk 6 above)
Comment (rf152) [reply src/Authentik/Provider.php:30]: done

File: src/Authentik/Provider.php (hunk 5 above)
Comment (rf152) [reply src/Authentik/Provider.php:13]: done

File: src/Authentik/composer.json (hunk 4 above)
Comment (rf152) [reply src/Authentik/composer.json:4]: Added

File: src/Authentik/Provider.php (hunk 8 above)
Comment (rf152) [reply src/Authentik/Provider.php:60]: done

File: src/Authentik/Provider.php (hunk 7 above)
Comment (lucasmichot) [reply src/Authentik/Provider.php:60]: Check other providers

File: src/Authentik/Provider.php [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+     * {@inheritdoc}
+     */
+    public static function additionalConfigKeys()
+    {
+        return ['base_url'];
+    }
+
+    protected function getBaseUrl()
+    {
+        $baseurl = $this->getConfgi('base_url');
```
Comment (lucasmichot) [inline src/Authentik/Provider.php:31]: typo

File: src/Authentik/Provider.php (hunk 7 above)
Comment (rf152) [reply src/Authentik/Provider.php]: Done

**Review by lucasmichot — APPROVED** [review]
Thx for the PR @rf152 
GTM @atymic

File: src/Authentik/Provider.php [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        return (new User())->setRaw($user)->map([
+            'email'              => $user['email'] ?? '',
+            'email_verified'     => $user['email_verified'] ?? null,
+            'name'               => $user['name'] ?? '',
+            'given_name'         => $user['given_name'] ?? '',
+            'family_name'        => $user['family_name'] ?? '',
+            'preferred_username' => $user['preferred_username'] ?? '',
+            'nickname'           => $user['nickname'] ?? '',
+            'groups'             => $user['groups'] ?? '',
```
Comment (atymic) [inline src/Authentik/Provider.php:86]: Is there a reason we default to empty string here, not null?
I think null makes mores sense - the value isn't present

**Review by atymic — COMMENTED** [review]
Just one minor change

File: src/Authentik/Provider.php (hunk 13 above)
Comment (AlexZorzi) [reply src/Authentik/Provider.php:86]: it was my suggestion, but yes, returning null would be better

**Review by atymic — APPROVED** [review]

**Comment by AlexZorzi** [comment]
@atymic how long does it take for the changes to be available on composer?

**Comment by atymic** [comment]
https://packagist.org/packages/socialiteproviders/authentik

**Comment by atymic** [comment]
Please also PR to the website repo

**Pull request merged**